proto/cosmos/base/reflection/v1beta1/reflection.proto

```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 2;
+}
+
+// InterfaceAcceptingMessageDescriptor describes a protobuf message which contains
+// an interface represented as a google.protobuf.Any
+message InterfaceAcceptingMessageDescriptor {
+  // fullname is the protobuf fullname of the type containing the interface
+  string fullname = 1;
+  // field_descriptor_names is a list of the protobuf name (not fullname) of the field
+  // which contains the interface as google.protobuf.Any (the interface is the same, but
+  // it can be in multiple fields of the same proto message)
+  repeated string field_descriptor_names = 2;
+}
+
+// ConfigurationDescriptor contains metadata information on the sdk.Config
+message ConfigurationDescriptor {
+  // bech32_account_address_prefix is the account address prefix
+  string bech32_account_address_prefix = 1;
+  // bech32_validator_address_prefix is the bech32 prefix for validator addresses
+  string bech32_validator_address_prefix = 2;
+  // bech32_consensus_address_prefix is the bech32 prefix for consensus nodes addresses
+  string bech32_consensus_address_prefix = 3;
+  // bech32_account_public_key_prefix is the bech32 prefix for accounts public keys
+  string bech32_account_public_key_prefix = 4;
+  // bech32_validator_public_key_prefix is the bech32 prefix for validators public keys
+  string bech32_validator_public_key_prefix = 5;
+  // bech32_consensus_public_key_prefix is the bech32 prefix for consensus nodes public keys
+  string bech32_consensus_public_key_prefix = 6;
+  // purpose is the BIP-0044 purpose code
+  uint32 purpose = 7;
+  // coin_type is the BIP-0044 coin type code
+  uint32 coin_type = 8;
+  // full_fundraiser_path is the BIP-0044 prefix (deprecated in favor of full_bip44_path)
+  string full_fundraiser_path = 9;
+  // full_bip44_path is the BIP-0044 prefix
+  string full_bip44_path = 10;
+}
+
+// MsgDescriptor describes a cosmos-sdk message that can be delivered with a transaction
+message MsgDescriptor {
+  // msg contains a descriptor of sdk.ServiceMsg or sdk.Msg
+  oneof msg {
+    // service_msg is used when the message is an sdk.ServiceMsg type
+    ServiceMsgDescriptor service_msg = 1;
+    // legacy_msg is used when the message is an sdk.Msg type
+    LegacyMsgDescriptor legacy_msg = 2;
+  }
+}
+
+// ServiceMsgDescriptor describes an sdk.ServiceMsg type
+message ServiceMsgDescriptor {
+  // fullname is the protobuf fullname of the given sdk.ServiceMsg
+  string fullname = 1;
+  // route is the sdk.ServiceMsg route, it is equal to type_url
+  string route = 2;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 3;
```

Comment:
These are all redundant as far as I know. We just need the type URL.

Reply:
The concrete name is required too for marshal options to get the correct message descriptor which is saved via protobuf.Fullname, type URL is custom to our codec implementation

Reply:
Can you give me an example of what all three of these would be for a Msg?

Reply:
Yes sure. For example when I resolve the ServiceMsg interface. I get the following type URL: `/cosmos.crisis.v1beta1.Msg/VerifyInvariant` 

Now if I am a reflection client, how do I know which is the protobuf fullname for this typeURL? The protobuf fullname is actually: `cosmos.crisis.v1beta1.MsgVerifyInvariant` so I can now query the gRPC reflection to give me the file which contains this symbol. 

Here you have an example of the app descriptor for simapp: https://pastebin.com/BxSyKXJk

Reply:
Well there's a distinction in ServiceMsg's between the method and the request type. What you're calling the full name is actually the request type name. I think we should have two fields for request type full name and response type full name for service msgs.

Reply:
Also ServiceMsg's do abuse a bit the type URL of Any, that's covered in ADR 031 as sort of a necessary "abuse" so this can happen. Hope it's clear enough what's actually happening.

Reply:
Sorry I didn't understand, so do you agree that we need both typeURL and protobuf fullname (the name the type is registered with in the protobuf registry, not our codec) so when we marshal the serviceMsg as anypb it will have the correct type URL, and the protobuf fullname so that we can build the correct message via descriptors?

> I think we should have two fields for request type full name and response type full name for service msgs.

I don't know how to fetch those, currently I don't think anywhere in the app we save the `Msg` server service descriptors?

Reply:
Side question: why in ADR-031 we preferred the custom type URL, instead of the `MsgServer` service descriptor method fullname?

Reply:
the configurator which holds `MsgServer` descriptors is defined at specific app level (not `baseapp` level), so the reflection server cannot make use of that. I've defined in the `ServiceMsgDescriptor` the `response_fullname` but as of now we cannot make use of that. After the configurator finds its place in `baseapp` we can fill the missing information.